library/common.h

```diff
@@ -195,4 +231,17 @@ inline void mbedtls_xor(unsigned char *r, const unsigned char *a, const unsigned
 #define MBEDTLS_UNLIKELY(x)     x
 #endif
 
+#if defined(__GNUC__) && !defined(__ARMCC_VERSION) && !defined(__clang__) \
+    && !defined(__llvm__) && !defined(__INTEL_COMPILER)
+/* Defined if the compiler really is gcc and not clang, etc */
+#define MBEDTLS_COMPILER_IS_GCC
+#endif
+
+/* If -Os is specified, override with -O2 for a given function */
+#if defined(MBEDTLS_COMPILER_IS_GCC) && defined(__OPTIMIZE_SIZE__)
+#define MBEDTLS_OPTIMIZE_ALWAYS __attribute__((optimize("-O2")))
```

Comment:
This name is misleading: it suggests to always optimize even if optimizations are disabled, but doesn't convey what it's optimizing for. But actually it preserves the no-optimization choice, and transforms one choice of optimization into another.
```suggestion
#define MBEDTLS_OPTIMIZE_ALWAYS_FOR_PERFORMANCE __attribute__((optimize("-O2")))
```

Reply:
I'd suggest `MBEDTLS_OPTIMIZE_FOR_PERFORMANCE` :)

Reply:
If we want it to be really ambiguous, I suppose it would have to be something like `MBEDTLS_OPTIMIZE_FOR_PERFORMANCE_WHEN_OPTIMIZING_AT_ALL`. Even I shirk at such a verbose name, as long as it conveys the idea that it's changing what's being optimized and doesn't strongly suggest that it's changing whether optimizing at all.